modules/publication/ajax/FileUpload.php

```diff
@@ -0,0 +1,877 @@
+    }
+    // title that gets inserted is run through htmlspecialchars()
+    // so need to query based on Processed title
+    $titleProc = htmlspecialchars($titleRaw);
+
+    // back end validation for title uniqueness constraint
+    $exists = $db->pselectOne(
+        "SELECT PublicationID " .
+        "FROM publication " .
+        "WHERE Title=:t",
+        array('t' => $titleProc)
+    );
+
+    if ($exists) {
+        showPublicationError('Submitted title already exists', 400);
+    }
+    $desc            = $_POST['description'] ?? null;
+    $leadInvest      = $_POST['leadInvestigator'] ?? null;
+    $leadInvestEmail = $_POST['leadInvestigatorEmail'] ?? null;
+
+    // check if lead investigator already exists in collaborator table
+    // use ID if exists, else insert
+    $leadInvID = $db->pselectOne(
+        'SELECT PublicationCollaboratorID '.
+        'FROM publication_collaborator '.
+        'WHERE Name = :n OR Email = :e',
+        array(
+         'n' => $leadInvest,
+         'e' => $leadInvestEmail,
+        )
+    );
+    if (empty($leadInvID)) {
+        $db->insert(
+            'publication_collaborator',
+            array(
+             'Name'  => $leadInvest,
+             'Email' => $leadInvestEmail,
+            )
+        );
+
+        $leadInvID = $db->getLastInsertId();
+    }
+    if (!isset($desc, $leadInvest, $leadInvestEmail)) {
+        showPublicationError('A mandatory field is missing!', 400);
+    }
+    // INSERT INTO publication ...
+    $uid   = $user->getId();
+    $today = date('Y-m-d');
+    // insert the titleRaw to avoid double escaping
+    $fields = array(
+               'UserID'             => $uid,
+               'Title'              => $titleRaw,
+               'Description'        => $desc,
+               'LeadInvestigatorID' => $leadInvID,
+               'DateProposed'       => $today,
+              );
+
+    $db->insert('publication', $fields);
+    $pubID = $db->getLastInsertId();
+
+    try {
+        // process files
+        processFiles($pubID);
+        // INSERT INTO publication_collaborator
+        insertCollaborators($pubID);
+        // INSERT INTO publication_users_edit_perm_rel
+        insertEditors($pubID);
+        // INSERT INTO publication_keyword
+        insertKeywords($pubID);
+        // INSERT INTO publication_parameter_type_rel
+        insertVOIs($pubID);
+    } catch (Exception $e) {
+        cleanup($pubID);
+        showPublicationError($e->getMessage(), 500);
+    }
+
+    notify($pubID, 'submission');
+}
+
+/**
+ * Function for storing files and inserting file meta data into database
+ *
+ * @param int $pubID ID of Publication in publication table
+ *
+ * @return void
+ */
+function processFiles($pubID) : void
+{
+    if (empty($_FILES)) {
+        return;
+    }
+    $db     = \Database::singleton();
+    $config = \NDB_Config::singleton();
+
+    $publicationPath = $config->getSetting('publication_uploads');
+
+    if (!is_dir($publicationPath)) {
```

Comment:
Should there be a check that the path ends with a slash so that line 150 doesn't break?

Reply:
added a check to append a trailing slash in case there is none in the configured value